File: typescript/bc-envelope/src/format-context.ts
import {
    type Cbor,
    type SummarizerResult,
    type Tag,
    TagsStore,
    type TagsStoreTrait,
    withTagsMut,
} from '@bc/dcbor';
import {
    TAG_EVENT,
    TAG_FUNCTION,
    TAG_KNOWN_VALUE,
    TAG_PARAMETER,
    TAG_REQUEST,
    TAG_RESPONSE,
    registerTagsIn as registerBCTagsIn,
} from '@bc/tags';
import { KNOWN_VALUES, type KnownValuesStore, KnownValue } from '@bc/known-values';
import {
    registerTags as registerComponentTags,
    registerTagsIn as registerComponentTagsIn,
} from '@bc/components';

import { flankedBy } from './string-utils.js';
import { GLOBAL_FUNCTIONS, FunctionsStore } from './functions-store.js';
import { GLOBAL_PARAMETERS, ParametersStore } from './parameters-store.js';
import {
    functionFromUntaggedCbor,
} from './function.js';
import {
    parameterFromUntaggedCbor,
} from './parameter.js';
import { Envelope } from './envelope.js';

export type FormatContextOpt =
    | { type: 'none' }
    | { type: 'global' }
    | { type: 'custom'; context: FormatContext };

export const FormatContextOpts = {
    none(): FormatContextOpt {
        return { type: 'none' };
    },
    global(): FormatContextOpt {
        return { type: 'global' };
    },
    custom(context: FormatContext): FormatContextOpt {
        return { type: 'custom', context };
    },
};

export class FormatContext implements TagsStoreTrait {
    readonly #tags: TagsStore;
    readonly #knownValues: KnownValuesStore;
    readonly #functions: FunctionsStore;
    readonly #parameters: ParametersStore;

    constructor(
        tags = new TagsStore(),
        knownValues: KnownValuesStore = KNOWN_VALUES,
        functions = GLOBAL_FUNCTIONS,
        parameters = GLOBAL_PARAMETERS,
    ) {
        this.#tags = tags;
        this.#knownValues = knownValues;
        this.#functions = functions;
        this.#parameters = parameters;
    }

    tags(): TagsStore {
        return this.#tags;
    }

    knownValues(): KnownValuesStore {
        return this.#knownValues;
    }

    functions(): FunctionsStore {
        return this.#functions;
    }

    parameters(): ParametersStore {
        return this.#parameters;
    }

    assignedNameForTag(tag: Tag): string | undefined {
        return this.#tags.assignedNameForTag(tag);
    }

    nameForTag(tag: Tag): string {
        return this.#tags.nameForTag(tag);
    }

    tagForName(name: string): Tag | undefined {
        return this.#tags.tagForName(name);
    }

    tagForValue(value: bigint): Tag | undefined {
        return this.#tags.tagForValue(value);
    }

    summarizer(tag: bigint) {
        return this.#tags.summarizer(tag);
    }

    nameForValue(value: bigint): string {
        return this.#tags.nameForValue(value);
    }
}

let globalContext: FormatContext | undefined;

export function getGlobalFormatContext(): FormatContext {
    if (globalContext === undefined) {
        // Ensure the global dCBOR tag store contains named bc tags for UR encoding.
        registerComponentTags();

        const tags = new TagsStore();
        const context = new FormatContext(tags, KNOWN_VALUES, GLOBAL_FUNCTIONS, GLOBAL_PARAMETERS);
        registerTagsIn(context);
        globalContext = context;
    }
    return globalContext;
}

export function withFormatContext<T>(action: (context: FormatContext) => T): T {
    return action(getGlobalFormatContext());
}

export function registerTagsIn(context: FormatContext): void {
    registerBCTagsIn(context.tags());
    registerComponentTagsIn(context.tags());

    const ok = (value: string): SummarizerResult => ({ ok: true, value });

    const knownValues = context.knownValues();
    context.tags().setSummarizer(TAG_KNOWN_VALUE, (untaggedCbor) => {
        const kv = KnownValue.fromUntaggedCbor(untaggedCbor);
        return ok(flankedBy(knownValues.name(kv), "'", "'"));
    });

    const functions = context.functions();
    context.tags().setSummarizer(TAG_FUNCTION, (untaggedCbor) => {
        const fn = functionFromUntaggedCbor(untaggedCbor);
        return ok(flankedBy(FunctionsStore.nameForFunction(fn, functions), '\u00AB', '\u00BB'));
    });

    const parameters = context.parameters();
    context.tags().setSummarizer(TAG_PARAMETER, (untaggedCbor) => {
        const param = parameterFromUntaggedCbor(untaggedCbor);
        return ok(flankedBy(ParametersStore.nameForParameter(param, parameters), '\u2770', '\u2771'));
    });

    const summarizeEnvelope = (untaggedCbor: Cbor, flat: boolean, wrapper: string): SummarizerResult => {
        const envelope = Envelope.newLeaf(untaggedCbor);
        const formatted = envelope.formatOpt(flat, FormatContextOpts.custom(context));
        return ok(`${wrapper}(${formatted})`);
    };

    context.tags().setSummarizer(TAG_REQUEST, (untaggedCbor, flat) => summarizeEnvelope(untaggedCbor, flat, 'request'));
    context.tags().setSummarizer(TAG_RESPONSE, (untaggedCbor, flat) => summarizeEnvelope(untaggedCbor, flat, 'response'));
    context.tags().setSummarizer(TAG_EVENT, (untaggedCbor, flat) => summarizeEnvelope(untaggedCbor, flat, 'event'));
}

export function registerTags(): void {
    withTagsMut((tagsStore) => {
        registerBCTagsIn(tagsStore);
        registerComponentTagsIn(tagsStore);
    });
    getGlobalFormatContext();
}

File: go/bcenvelope/format_context.go
package bcenvelope

import (
	"sync"

	bccomponents "github.com/nickel-blockchaincommons/bccomponents-go"
	bctags "github.com/nickel-blockchaincommons/bctags-go"
	dcbor "github.com/nickel-blockchaincommons/dcbor-go"
	knownvalues "github.com/nickel-blockchaincommons/knownvalues-go"
)

// FormatContextMode determines which format context to use.
type FormatContextMode int

const (
	// FormatContextNone uses no format context.
	FormatContextNone FormatContextMode = iota
	// FormatContextGlobal uses the process-global format context.
	FormatContextGlobal
	// FormatContextCustom uses a caller-provided format context.
	FormatContextCustom
)

// FormatContextOpt configures which format context source to use when formatting.
type FormatContextOpt struct {
	Mode    FormatContextMode
	Context *FormatContext
}

// FormatOptNone selects no format context.
func FormatOptNone() FormatContextOpt {
	return FormatContextOpt{Mode: FormatContextNone}
}

// FormatOptGlobal selects the process-global format context.
func FormatOptGlobal() FormatContextOpt {
	return FormatContextOpt{Mode: FormatContextGlobal}
}

// FormatOptCustom selects a caller-provided format context.
func FormatOptCustom(ctx *FormatContext) FormatContextOpt {
	return FormatContextOpt{Mode: FormatContextCustom, Context: ctx}
}

// ToTagsStoreOpt converts a FormatContextOpt to a dcbor.TagsStoreOpt.
func (o FormatContextOpt) ToTagsStoreOpt() dcbor.TagsStoreOpt {
	switch o.Mode {
	case FormatContextNone:
		return dcbor.TagsNone()
	case FormatContextGlobal:
		return dcbor.TagsGlobal()
	case FormatContextCustom:
		if o.Context != nil {
			return dcbor.TagsCustom(o.Context.Tags())
		}
		return dcbor.TagsNone()
	default:
		return dcbor.TagsNone()
	}
}

// FormatContext provides information about CBOR tags, known values, functions,
// and parameters that are used to annotate the output of envelope formatting functions.
type FormatContext struct {
	tags        *dcbor.TagsStore
	knownValues *knownvalues.KnownValuesStore
	functions   *FunctionsStore
	parameters  *ParametersStore
}

// NewFormatContext creates a new format context with the specified components.
// Any nil component is replaced with its default.
func NewFormatContext(
	tags *dcbor.TagsStore,
	kv *knownvalues.KnownValuesStore,
	functions *FunctionsStore,
	parameters *ParametersStore,
) *FormatContext {
	if tags == nil {
		tags = dcbor.NewTagsStore(nil)
	}
	if kv == nil {
		kv = knownvalues.NewKnownValuesStore()
	}
	if functions == nil {
		functions = NewFunctionsStore()
	}
	if parameters == nil {
		parameters = NewParametersStore()
	}
	return &FormatContext{
		tags:        tags,
		knownValues: kv,
		functions:   functions,
		parameters:  parameters,
	}
}

// DefaultFormatContext creates a format context with default (empty) components.
func DefaultFormatContext() *FormatContext {
	return NewFormatContext(nil, nil, nil, nil)
}

// Tags returns the CBOR tags registry.
func (c *FormatContext) Tags() *dcbor.TagsStore { return c.tags }

// KnownValues returns the known values registry.
func (c *FormatContext) KnownValues() *knownvalues.KnownValuesStore { return c.knownValues }

// Functions returns the functions registry.
func (c *FormatContext) Functions() *FunctionsStore { return c.functions }

// Parameters returns the parameters registry.
func (c *FormatContext) Parameters() *ParametersStore { return c.parameters }

// Clone returns an independent copy of the format context.
func (c *FormatContext) Clone() *FormatContext {
	if c == nil {
		return DefaultFormatContext()
	}
	// Tags store is shared (not deep cloned) since TagsStore has no Clone
	// in the Go dcbor package; callers can register additional tags.
	return &FormatContext{
		tags:        c.tags,
		knownValues: c.knownValues.Clone(),
		functions:   c.functions.Clone(),
		parameters:  c.parameters.Clone(),
	}
}

// TagsResolver interface delegation methods.

// AssignedNameForTag returns the registered name for the tag value, if present.
func (c *FormatContext) AssignedNameForTag(tag dcbor.Tag) (string, bool) {
	return c.tags.AssignedNameForTag(tag)
}

// NameForTag returns the assigned name for a tag, or numeric fallback text.
func (c *FormatContext) NameForTag(tag dcbor.Tag) string {
	return c.tags.NameForTag(tag)
}

// TagForValue looks up a tag definition by numeric value.
func (c *FormatContext) TagForValue(value dcbor.TagValue) (dcbor.Tag, bool) {
	return c.tags.TagForValue(value)
}

// TagForName looks up a tag definition by assigned name.
func (c *FormatContext) TagForName(name string) (dcbor.Tag, bool) {
	return c.tags.TagForName(name)
}

// NameForValue returns the assigned name for a value, or numeric fallback text.
func (c *FormatContext) NameForValue(value dcbor.TagValue) string {
	return c.tags.NameForValue(value)
}

// Summarizer looks up the registered summarizer for a tag value.
func (c *FormatContext) Summarizer(tag dcbor.TagValue) (dcbor.CBORSummarizer, bool) {
	return c.tags.Summarizer(tag)
}

// Verify FormatContext implements dcbor.TagsResolver.
var _ dcbor.TagsResolver = (*FormatContext)(nil)

// Global format context singleton.
var (
	globalFormatContext     *FormatContext
	globalFormatContextOnce sync.Once
	globalFormatContextMu   sync.Mutex
)

func initGlobalFormatContext() {
	globalFormatContextOnce.Do(func() {
		bccomponents.RegisterTags()
		tagsStore := dcbor.GlobalTags.Get()
		kvStore := knownvalues.KnownValues.Get()
		functionsStore := GlobalFunctions()
		parametersStore := GlobalParameters()
		globalFormatContext = NewFormatContext(tagsStore, kvStore, functionsStore, parametersStore)
	})
}

// WithFormatContext provides read access to the global format context under lock.
func WithFormatContext[T any](action func(*FormatContext) T) T {
	initGlobalFormatContext()
	globalFormatContextMu.Lock()
	defer globalFormatContextMu.Unlock()
	return action(globalFormatContext)
}

// UpdateFormatContext provides read-write access to the global format context under lock.
func UpdateFormatContext(action func(*FormatContext)) {
	initGlobalFormatContext()
	globalFormatContextMu.Lock()
	defer globalFormatContextMu.Unlock()
	action(globalFormatContext)
}

// RegisterTagsIn registers standard tags and summarizers in a format context.
func RegisterTagsIn(context *FormatContext) {
	bccomponents.RegisterTagsIn(context.Tags())

	// Known value summarizer
	kvStore := context.KnownValues().Clone()
	context.Tags().SetSummarizer(bctags.TagKnownValue, func(untaggedCBOR dcbor.CBOR, _ bool) (string, error) {
		kv, err := knownvalues.DecodeKnownValue(untaggedCBOR)
		if err != nil {
			return "", err
		}
		return "'" + kvStore.Name(kv) + "'", nil
	})

	// Function summarizer
	fStore := context.Functions().Clone()
	context.Tags().SetSummarizer(bctags.TagFunction, func(untaggedCBOR dcbor.CBOR, _ bool) (string, error) {
		f, err := DecodeFunction(untaggedCBOR)
		if err != nil {
			return "", err
		}
		return "\u00ab" + NameForFunction(f, fStore) + "\u00bb", nil
	})

	// Parameter summarizer
	pStore := context.Parameters().Clone()
	context.Tags().SetSummarizer(bctags.TagParameter, func(untaggedCBOR dcbor.CBOR, _ bool) (string, error) {
		p, err := DecodeParameter(untaggedCBOR)
		if err != nil {
			return "", err
		}
		return "\u2770" + NameForParameter(p, pStore) + "\u2771", nil
	})

	// Request summarizer
	clonedCtxReq := context.Clone()
	context.Tags().SetSummarizer(bctags.TagRequest, func(untaggedCBOR dcbor.CBOR, flat bool) (string, error) {
		e := EnvelopeFromCBOR(untaggedCBOR)
		opts := defaultEnvelopeFormatOpts()
		opts.flat = flat
		opts.context = FormatOptCustom(clonedCtxReq)
		return "request(" + e.FormatOpt(opts) + ")", nil
	})

	// Response summarizer
	clonedCtxResp := context.Clone()
	context.Tags().SetSummarizer(bctags.TagResponse, func(untaggedCBOR dcbor.CBOR, flat bool) (string, error) {
		e := EnvelopeFromCBOR(untaggedCBOR)
		opts := defaultEnvelopeFormatOpts()
		opts.flat = flat
		opts.context = FormatOptCustom(clonedCtxResp)
		return "response(" + e.FormatOpt(opts) + ")", nil
	})

	// Event summarizer
	clonedCtxEvent := context.Clone()
	context.Tags().SetSummarizer(bctags.TagEvent, func(untaggedCBOR dcbor.CBOR, flat bool) (string, error) {
		e := EnvelopeFromCBOR(untaggedCBOR)
		opts := defaultEnvelopeFormatOpts()
		opts.flat = flat
		opts.context = FormatOptCustom(clonedCtxEvent)
		return "event(" + e.FormatOpt(opts) + ")", nil
	})
}

// RegisterTags registers standard tags in the global format context.
func RegisterTags() {
	UpdateFormatContext(func(context *FormatContext) {
		RegisterTagsIn(context)
	})
}
